### PR TITLE
add pattern for matching Array[GeneratedMessage] 

### PR DESCRIPTION
**Pull request by feixie79**
when try to convert some old generated message with repeated type returned from method getFieldByNumber as Array[GeneratedMessage], I got JsonFormatException due to missing matching of Array.

therefore I add pattern for matching Array[GeneratedMessage]  
and convert it implicitly to Iterable by serialization

### CODE REVIEW

**Comment by thesamet** [comment]
Can you add a test case that demonstrates the exception that would be thrown prior to the fix?

**Comment by feixie79** [comment]
> Can you add a test case that demonstrates the exception that would be thrown prior to the fix?

I added a test case and some generated code with a small modification in the method getFieldByNumber according to our internal code to demo the exception. and a second commit to fix this.

**Comment by thesamet** [comment]
@feixie79 Is the fix in JsonFormat needed due to a local modification of the code generator? If so, let's think how to make JsonFormat extensible so you can also have a locally modified JsonFormat. If this is the code that is being emitted from ScalaPB, can you put in the actual proto in `src/test/protobuf`, instead of the generated code?

**Comment by feixie79** [comment]
> @feixie79 Is the fix in JsonFormat needed due to a local modification of the code generator? If so, let's think how to make JsonFormat extensible so you can also have a locally modified JsonFormat. If this is the code that is being emitted from ScalaPB, can you put in the actual proto in `src/test/protobuf`, instead of the generated code?

I finally made it clear that the problem is happening when we put the colleciton_type option on the proto to force the type to immutable.Seq and generate the case class via some older version like 0.10.2. 
I made a commit to demo this.

Actually this problem seems solved by newer scalapb verison at least since 0.10.11. But unfortunately our organization could not easily upgrade to the newer scalapb version. 
Do you think it is a good idea to add this fix to the new scalapb-json4s library to provide some kind of backward compatibility to use the JsonFormat also for old generated code?

**Comment by thesamet** [comment]
> Do you think it is a good idea to add this fix to the new scalapb-json4s library to provide some kind of backward compatibility to use the JsonFormat also for old generated code?

Yes, I'd like to have the fix rebased for the 0.10.x (sounds like this is where you actually need it to go), and for the master branch, only the test case portion, without the fix.

File: src/test/protobuf/oldRepeatables.proto [hunk 1]
```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
```
Comment (thesamet) [inline src/test/protobuf/oldRepeatables.proto:1]: can this be moved to `src/test/protobuf/custom_collection.proto`, and have the collection_type applied to only the field, not the whole file?

File: src/test/scala/scalapb/json4s/OldRepeatablesSpec.scala [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package scalapb.json4s
```
Comment (thesamet) [inline src/test/scala/scalapb/json4s/OldRepeatablesSpec.scala:1]: Please rename file and test to `CustomCollectionSpec`

File: src/test/protobuf/oldRepeatables.proto [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+import "scalapb/scalapb.proto";
+
+option (scalapb.options) = {
+  collection_type: "scala.collection.immutable.Seq"
+};
+
+import "google/protobuf/wrappers.proto";
+
+message OldRepeatablesTest {
```
Comment (thesamet) [inline src/test/protobuf/oldRepeatables.proto:13]: CustomCollectionTest

**Comment by thesamet** [comment]
Thanks for investigating further the problem, making it reproducible and identifying the versions impacted!

File: src/main/scala/scalapb/json4s/JsonFormat.scala [hunk 4]
```diff
@@ -298,46 +296,51 @@ class Printer private (config: Printer.PrinterConfig) {
...
-                )
-              }
-              key -> value
-            }.toSeq: _*)
-          )
-        } else {
-          b += JField(name, JArray(xs.map(toJson).toList))
-        }
+        serializeIterable(xs)
+      /*      case array: Array[GeneratedMessage] =>
```
Comment (thesamet) [inline src/main/scala/scalapb/json4s/JsonFormat.scala:300]: Add comment:
// In early versions of ScalaPB 0.10.x, using a custom collection of type immutable.Seq would result in an array type here leading to an exception without this case. This case is not needed for ScalaPB 0.10.11 and 0.11.0. [link to github issue]

**Comment by feixie79** [comment]
> > Do you think it is a good idea to add this fix to the new scalapb-json4s library to provide some kind of backward compatibility to use the JsonFormat also for old generated code?
> 
> Yes, I'd like to have the fix rebased for the 0.10.x (sounds like this is where you actually need it to go), and for the master branch, only the test case portion, without the fix.

I created two new PR accordingly to master and 0.10.x which clean commits. I will close this one for now.

**Pull request closed without merging**